server/athenian/api/controllers/features/github/pull_request_metrics.py

```diff
@@ -654,3 +654,53 @@ def _analyze(self, facts: pd.DataFrame, min_time: datetime, max_time: datetime,
         sizes = facts["size"].values.astype(object)
         sizes[self._calcs[0].peek == np.array(None)] = None
         return sizes
+
+
+class StagePendingCountCalculator(WithoutQuantilesMixin, SumMetricCalculator[int]):
+
+    dtype = int
+    stage_name = None
+
+    def _analyze(self, facts: np.ndarray, min_time: datetime, max_time: datetime,
+                 **kwargs) -> np.ndarray:
+        merged_mask, approved_mask, frr_mask = (
+            (v := facts[c].values) == v
+            for c in ("merged", "approved", "first_review_request"))
+
+        stage_indexes = {}
+        other = ~facts["done"].values
+        stage_indexes["release"] = np.where(merged_mask & other)[0]
+        other &= ~merged_mask
+        stage_indexes["merge"] = np.where(approved_mask & other)[0]
+        other &= ~approved_mask
+        stage_indexes["review"] = np.where(frr_mask & other)[0]
+        other &= ~frr_mask
+        stage_indexes["wip"] = np.where(other)[0]
```

Comment:
This is suboptimal. Each child class will run the same code. There are two options that I see:

1. Composition instead of inheritance. Declare this analysis as a dependency for each of the four Counter-s using `deps` (e.g., like in `WorkInProgressCounter`) and run that dependency manually in `_calc_stage_timings`.
2. Do only the required calculations in each of the four Counter-s.

Reply:
Ok, now this computation is done only once, PTAL if it fits you as well. Applying option 2 wouldn't have helped much since those calculations are done in a chain, so each counter would have had to re-do some computation anyways. One alternative would have been to make them a dependency chain but seems more complex than it should.